File: solutions/lua/binary-search/1/binary-search.lua
return function(array, target)
  local lower_bound = 1
  local upper_bound = #array

  while (lower_bound <= upper_bound) do
    local middle_index = math.floor(lower_bound / 2 + upper_bound / 2)
    local middle_item  = array[middle_index]

    if (middle_item == target) then return middle_index end

    if (middle_item < target) then
      lower_bound = middle_index + 1
    else
      upper_bound = middle_index - 1
    end
  end

  return -1
end

File: solutions/javascript/binary-search/1/binary-search.js
export const find = (items, searchItem) => {
  let lowerBound = 0;
  let upperBound = items.length - 1;

  while (lowerBound <= upperBound) {
    // IMPORTANT: Avoid addition overflow by distributing the half.
    const middleIndex = Math.floor(lowerBound / 2 + upperBound / 2);
    const middleItem  = items[middleIndex];

    if (middleItem === searchItem) return middleIndex;

    if (middleItem < searchItem) {
      lowerBound = middleIndex + 1;
    }
    else {
      upperBound = middleIndex - 1;
    }
  }

  throw new Error('Value not in array');
};
